### ISSUE TITLE
JDK-8220012: Accordion control holds reference to child pane after it is removed

### ISSUE THREAD

**Issue by fthevenet** [javafxports/openjdk-jfx#391]
The Accordion control keeps a reference to its first TitledPane even after all panes have been removed, which prevents it to be collected, alongside everything attached to that pane.

The reason for that is the skin for the control explicitly  keeps a reference to it named firstTitledPane, which is kept in sync with the list of panes in the control through the use of a ListChangeListener, but it doesn't appear to ever sever the reference after the last TitlePane has been removed:

https://github.com/javafxports/openjdk-jfx/blob/3179dbd96653c22f329d75e7c7074612aef337b3/modules/javafx.controls/src/main/java/javafx/scene/control/skin/AccordionSkin.java#L101-L118

Unless I'm missing something, simply setting firstTitledPane to null if control.getPanes().isEmpty() is true should do the trick.

**Comment by fthevenet** [comment]
I have filed a bug in the JBS for this issue: [JDK-8220012](https://bugs.openjdk.java.net/browse/JDK-8220012)

**Comment by kevinrushforth** [comment]
Thank you for filing the JBS bug.

If you are interested in contributing a fix for this issue, please read [CONTRIBUTING.md](https://github.com/javafxports/openjdk-jfx/blob/develop/.github/CONTRIBUTING.md).

**Comment by kevinrushforth** [comment]
Also, we will need a test case that reproduces the bug.

**Comment by fthevenet** [comment]
I've submitted the following to the JBS bug:

> Steps to reproduce the issue:
> - Create a scene with an Accordion control
> - Start monitoring the heap usage.
> - Add a TitledPane to the Accordion, and to that pane any number of child controls susceptible to have noticeable impact on the heap usage (in the application where the issue was spotted, it consists of several XYCharts with several thousands of samples each, for instance)
> - Take a note of the increase in heap usage.
> - Remove the TitledPane from the Accordion
> - Force a GC
> - Notice that the old gen usage on the heap does not significantly decrease
> - Add a new TitledPane, this time with no children,  to the accordion
> - Force a GC
> - Notice that now the heap usage has decrease significantly, as the reference to the "big" titled pane was replace by one the the "small" one, and the former could therefore be collected.
> 
> Alternatively, using a Java profiler capable of tracing paths of strongly referenced objects to GC roots can also be used to illustrate the issue more accurately (I used Yourkit to identify the bug in the first place, for instance.)
>

**Comment by kevinrushforth** [comment]
Please provide a small, standalone test program. Additionally, as noted in PR #395, we will need an automated unit test that can verify this fix. For reference, you can see the test case provided in PR #342, [InitialNodesMemoryLeakTest.java](https://github.com/javafxports/openjdk-jfx/blob/develop/tests/system/src/test/java/test/javafx/scene/InitialNodesMemoryLeakTest.java).

**Comment by fthevenet** [comment]
Thanks for pointing out PR #342; I was in fact trying to figure out a robust way to test that kind of thing, but had not come up with anything really satisfying so far...
I will review it and write the unit test.

Meanwhile, I've submitted a signed OCA form.

**Comment by fthevenet** [comment]
I've added a unit test to PR #395 

Please also find below a standalone test program that illustrates the issue:
```
package AccordionTitlePaneLeakTest;

import javafx.application.Application;
import javafx.scene.Scene;
import javafx.scene.control.Accordion;
import javafx.scene.control.TitledPane;
import javafx.scene.layout.StackPane;
import javafx.stage.Stage;

import java.lang.ref.WeakReference;

public class AccordionTitlePaneLeakTest extends Application {
    public void start(Stage stage) throws Exception {
        Accordion accordion = new Accordion();
        StackPane root = new StackPane(accordion);
        stage.setScene(new Scene(root));
        stage.show();
        TitledPane pane = new TitledPane();
        accordion.getPanes().add(pane);
        WeakReference<TitledPane> weakRefToPane = new WeakReference<>(pane);
        pane = null;
        accordion.getPanes().clear();
        // At that point we expect 'pane' to no longer be strongly referenced,
        // and that the code below will cause it to be collected.
        for (int i = 0; i < 10; i++) {
            System.gc();
            System.runFinalization();
            if (weakRefToPane.get() == null) {
                break;
            }
            Thread.sleep(500);
        }
        // The following assertion will fail if 'pane' is still strongly referenced.
       assert weakRefToPane.get() == null :"Couldn't collect TitledPane";
    }

    public static void main(String[] args) {
        launch(args);
    }
}
```

**Comment by fthevenet** [comment]
I've send a signed copy of the OCA form by email two weeks ago (march 2nd) but I still haven't heard anything; do you know where I could inquire, or should I just wait some more?

**Comment by fthevenet** [comment]
It's been another full week and I still haven't heard back from Oracle regarding the contribution agreement.

I've resent it, in case it got lost somehow the first time around... Meanwhile I have to say the whole experience is a tad frustrating.

**Comment by fthevenet** [comment]
Fixed in b5e56d9b867b2c86186bc7dc0a68e021c6cb0a9f
